lib/transaction/priority_queue.go

```diff
@@ -139,6 +142,30 @@ func (spq *PriorityQueue) Push(txn *ValidTransaction) (common.Hash, error) {
 	return hash, nil
 }
 
+// PopChannel returns a *ValedTransaction channel to be signalled
+// when the next Push() is called on the queue.
+func (spq *PriorityQueue) PopChannel(timer *time.Timer) (tx chan *ValidTransaction) {
+	popChannel := make(chan *ValidTransaction)
+	go func() {
+		for {
+			select {
+			case <-timer.C:
+				close(popChannel)
+			default:
+				txn := spq.Pop()
+
+				if txn != nil {
+					popChannel <- txn
+				} else {
+					time.Sleep(spq.sleepTime)
```

Comment:
I think the point of the PR is to solve the unrestricted loop, the audit suggested adding a sleep into the loop to solve this.  I created PopChannel function on PriorityQueue so that notification and time out logic is contained within that function.  The sleep is used when the queue is empty to let that thread rest.

Reply:
The solution proposed in the audit has the problem that what value should be the poll period? What if it's smaller/larger than the slot time? @timwu20 any thoughts on that?

Also if we still want to do time based, you can simplify all this to be something like https://github.com/ChainSafe/gossamer/pull/2829 and not have this extra PopChannel method and not expose channels at all.

Reply:
> The solution proposed in the audit has the problem that we will eventually miss some extrinsics whilst we sleep the last time before the slot ends.

@qdm12 if you think this is a problem then state this explicitly.  Don't conflate this with what the auditors have recommended. 

I paired with @edwardmack on this, and designed the PopChannel interface to abstract away the idea of reading all `ValidTransaction` instances from the priority queue over a given time.  

> The solution proposed in the audit has the problem that we will eventually miss some extrinsics whilst we sleep the last time before the slot ends. This is true even if the sleep is very short. Time based is just not the right approach imo.

The existing code reserved 1/3 of the time for finalisation.  "missed" extrinsics during this duration of time will be added to the priority queue and included in the next block.  This is the already existing behaviour on the base branch.  So what approach would be the right approach in your opinion?

Reply:
We had an implementation fully event driven (no timer to retry polling) from commit https://github.com/ChainSafe/gossamer/pull/2632/commits/7ae3a2185c9a12bd30ef7ae97e3512f3fe96b3d6

It got dropped later (not sure which commit) in favor of the current periodic polling. Anyway if we want to keep the periodic polling, why not just do what the first commit was doing: https://github.com/ChainSafe/gossamer/pull/2632/commits/b6ecc25bc8cb2932d9a268064eb4a2391ffe8518 🤔

Reply:
> We had an implementation fully event driven (no timer to retry polling) from commit https://github.com/ChainSafe/gossamer/commit/7ae3a2185c9a12bd30ef7ae97e3512f3fe96b3d6

Again I had paired with @edwardmack and designed the `PopChannel` idea together.  We wanted to simplify the calling code to not have to worry about timers, or how/when to pop form the queue or create some sort of `pushWatcher`, which can all be contained within the bounded context of the priority queue.

I'm not sure how beneficial it is to rehash old commits, but my feedback around that commit specifically was that the handling of `pushWatchers` is quite cumbersome.  Additionally, there's a chance the priority queue could panic as a result of trying to close a `nil` channel or an already closed channel.

We could've taken the `pushWatcher` component and implemented it into the `popChannel`, but it seemed like overkill, since the calling code is now interfacing with a separate producing goroutine, which would sleep on its own.  And it can still be added as an enhancement.

> It got dropped later (not sure which commit) in favor of the current periodic polling. Anyway if we want to keep the periodic polling, why not just do what the first commit was doing: https://github.com/ChainSafe/gossamer/commit/b6ecc25bc8cb2932d9a268064eb4a2391ffe8518 🤔

Again, we wanted to simplify the calling code to be able to just get all the messages for a supplied amount of time.  If we just added the sleep, there's a chance that that we would exit late if the priority queue was empty which initiated a sleep, at the same time the timer had elapsed.